File: server/Requests.py

from abc import ABC, abstractmethod
import struct
from constants import *

class RequestHeader:
    client_id: bytes
    version: int
    code: int
    payload_size: int

    def __init__(self, clientid:bytes, version:int, code:int, payloadsize:int):
        self.client_id = clientid
        self.version = version
        self.code = code
        self.payload_size = payloadsize

class Request(ABC):
    header:RequestHeader

    @staticmethod
    @abstractmethod
    def create_request_from_payload(header:RequestHeader, payload:bytes):
        pass

class Registration(Request):
    name:str

    @staticmethod
    def create_request_from_payload(header:RequestHeader, payload:bytes) -> Request:
        client_name = struct.unpack(f"{FieldsSizes.CLIENT_NAME}s", payload)[0]

        return Registration(header, client_name)

    def __init__(self, header:RequestHeader, client_name):
        self.header = header
        self.name = client_name

class PublicKeyTransfer(Request):
    name:str
    public_key:str

    @staticmethod
    def create_request_from_payload(header:RequestHeader, payload:bytes) -> Request:
        client_name, public_key = struct.unpack(f"{FieldsSizes.CLIENT_NAME}s{FieldsSizes.PUBLIC_KEY}s", payload)
        return PublicKeyTransfer(header, client_name, public_key)

    def __init__(self, header:RequestHeader, client_name, public_key):
        self.header = header
        self.client_name = client_name
        self.public_key = public_key

class Relogin(Request):
    name:str

    @staticmethod
    def create_request_from_payload(header:RequestHeader, payload:bytes) -> Request:
        client_name = struct.unpack(f"{FieldsSizes.CLIENT_NAME}s", payload)[0]

        return Relogin(header, client_name)

    def __init__(self, header:RequestHeader, client_name):
        self.header = header
        self.client_name = client_name

class FileTransfer(Request):
    content_size:int
    original_file_size:int
    packet_number:int
    total_packets:int
    file_name:str
    message_content:bytes

    @staticmethod
    def create_request_from_payload(header:RequestHeader, payload:bytes) -> Request:
        offset = FieldsSizes.CONTENT_SIZE + FieldsSizes.ORIGINAL_CONTENT_SIZE + FieldsSizes.PACKET_NUMBER + FieldsSizes.TOTAL_PACKETS

        content_size, original_size, packet_number, total_packets = struct.unpack("<IIHH", payload[0:offset])

        payload = payload[offset:]

        offset = FieldsSizes.FILE_NAME

        file_name = struct.unpack(f"{FieldsSizes.FILE_NAME}s", payload[0:offset])[0].decode()

        payload = payload[offset:]

        # Todo: check this
        #  message_content = struct.unpack("<", payload)[0]
        message_content = payload

        return FileTransfer(header, content_size, original_size, packet_number, total_packets, file_name, message_content)

    def __init__(self, header:RequestHeader, content_size, original_file_size, packet_number, total_packets, file_name, message_content):
        self.header = header
        self.content_size = content_size
        self.original_file_size = original_file_size
        self.packet_number = packet_number
        self.total_packets = total_packets
        self.file_name = file_name
        self.message_content = message_content

class ValidCRC(Request):
    file_name:str

    @staticmethod
    def create_request_from_payload(header:RequestHeader, payload:bytes) -> Request:
        file_name = struct.unpack(f"{FieldsSizes.FILE_NAME}s", payload)[0]

        return ValidCRC(header, file_name)
    def __init__(self, header:RequestHeader, file_name):
        self.header = header
        self.file_name = file_name

class InvalidCRC(Request):
    file_name:str

    @staticmethod
    def create_request_from_payload(header:RequestHeader, payload:bytes) -> Request:
        file_name = struct.unpack(f"{FieldsSizes.FILE_NAME}s", payload)[0]

        return ValidCRC(header, file_name)

    def __init__(self, header:RequestHeader, file_name):
        self.header = header
        self.file_name = file_name

class InvalidCRCFourthTime(Request):
    file_name:str

    @staticmethod
    def create_request_from_payload(header:RequestHeader, payload:bytes) -> Request:
        file_name = struct.unpack(f"{FieldsSizes.FILE_NAME}s", payload)[0]

        return ValidCRC(header, file_name)

    def __init__(self, header:RequestHeader, file_name):
        self.header = header
        self.file_name = file_name

def parse_request(conn) -> Request:
    Request_Codes_To_Handlers = {
        RequestCodes.REGISTRATION: Registration.create_request_from_payload,
        RequestCodes.PUBLIC_KEY_TRANSFER: PublicKeyTransfer.create_request_from_payload,
        RequestCodes.RELOGIN: Relogin.create_request_from_payload,
        RequestCodes.FILETRANSFER: FileTransfer.create_request_from_payload,
        RequestCodes.VALIDCRC: ValidCRC.create_request_from_payload,
        RequestCodes.INVALIDCRC: InvalidCRC.create_request_from_payload,
        RequestCodes.INVALIDCRCFOURTHTIME: InvalidCRCFourthTime.create_request_from_payload
    }

    try:
        request_bytes = conn.recv(FieldsSizes.HEADER)  # Get the header

        client_id, version, code, payload_size = struct.unpack(f"<{FieldsSizes.CLIENT_ID}sBHL", request_bytes)

        # if code not in Request_Codes_To_Handlers:
        # TODO: output error, invalid code

        # if payload_size != desired_payload_size:
        # TODO: output error, invalid payload size

        header = RequestHeader(client_id, version, code, payload_size)

        payload_bytes: bytes = conn.recv(payload_size)

        return Request_Codes_To_Handlers[code](header, payload_bytes)

    except Exception as e:
        print(e)
        return None # type: ignore

File: client/Requests.cpp
#include <iostream>
#include <algorithm>
#include <string.h>
#include "Requests.h"
#include "Utilities.h"
#include "User.h"

RequestHeader& RequestHeader::operator=(const RequestHeader& other) {
	if(this == &other)
		return *this;

	std::copy(other.client_id, other.client_id + Constants::Sizes_In_Bytes::CLIENT_ID, client_id);
	version = other.version;
	code = other.code;
	payload_size = other.payload_size;

	return *this;
}

void RequestHeader::init(const char client_id[], uint16_t code, uint32_t payload_size) {
	std::copy(client_id, client_id + Constants::Sizes_In_Bytes::CLIENT_ID, this->client_id);
	this->version = Constants::CLIENT_VERSION;
	this->code = code;
	this->payload_size = payload_size;
}

std::vector<uint8_t> RequestHeader::pack() {
	std::vector<uint8_t> result {};

	for(int i = 0 ; i < Constants::Sizes_In_Bytes::CLIENT_ID ; i++)
		result.push_back(client_id[i]);

	uint16_t _code = code;
	uint32_t _payload_size = payload_size;

	result.push_back(version);

	// Send it in little endian

	for(int i = sizeof(_code) ; i >= 1 ; i-- )
		result.push_back(Hex::get_byte(_code, i));

	for(int i = sizeof(_payload_size); i >= 1 ; i--)
		result.push_back(Hex::get_byte(_payload_size, i));

	return result;
}

Registration::Registration(const User& u) {
	this->header.init(Constants::EMPTY_NAME, Constants::Requests::codes::Registration, Constants::Requests::payload_sizes::Registration);

	std::copy(u.name, u.name + Constants::Sizes_In_Bytes::CLIENT_NAME, this->client_name);
}

std::vector<uint8_t> Registration::pack() {
	std::vector<uint8_t> res = header.pack();
	
	for(int i = 0 ; i < Constants::Sizes_In_Bytes::CLIENT_NAME ; i++)
		res.push_back(client_name[i]);

	return res;
}

PublicKeyTransfer::PublicKeyTransfer(const User& u) {
	this->header.init(u.uuid, Constants::Requests::codes::PublicKeyTransfer, Constants::Requests::payload_sizes::PublicKeyTransfer);

	std::copy(u.name, u.name + Constants::Sizes_In_Bytes::CLIENT_NAME, this->client_name);

	string pub_key = u.rsa_object->getPublicKey();

	copy_from_string_to_array(this->public_key, Constants::Sizes_In_Bytes::PUBLIC_KEY, pub_key);
}

std::vector<uint8_t> PublicKeyTransfer::pack() {
	std::vector<uint8_t> res = header.pack();

	for(int i = 0; i < Constants::Sizes_In_Bytes::CLIENT_NAME ; i++)
		res.push_back(client_name[i]);


	for(int i = 0 ; i < Constants::Sizes_In_Bytes::PUBLIC_KEY ; i++)
		res.push_back(public_key[i]);

	return res;
}

Relogin::Relogin(const User& u) {
	this->header.init(u.uuid, Constants::Requests::codes::Relogin, Constants::Requests::payload_sizes::Relogin);
	std::copy(u.name, u.name + Constants::Sizes_In_Bytes::CLIENT_NAME, this->client_name);
}

std::vector<uint8_t> Relogin::pack() {
	std::vector<uint8_t> res = header.pack();

	for(int i = 0; i < Constants::Sizes_In_Bytes::CLIENT_NAME; i++)
		res.push_back(client_name[i]);

	return res;
}

FileTransfer::FileTransfer(const User& u, uint32_t content_size, uint32_t original_file_size, uint16_t packet_number, uint16_t total_packets) {

	this->header.init(u.uuid, Constants::Requests::codes::FileTransfer, Constants::Requests::payload_sizes::FileTransfer);
	this->content_size = content_size;
	this->original_file_size = original_file_size;
	this->packet_number = packet_number;
	this->total_packets = total_packets;

	std::copy(u.file_name, u.file_name + Constants::Sizes_In_Bytes::FILE_NAME, this->file_name);
}

std::vector<uint8_t> FileTransfer::pack() {
	std::vector<uint8_t> res = header.pack();
	
	// Send it in little endian
	for(int i = sizeof(content_size); i >= 1 ; i--)
		res.push_back(Hex::get_byte(content_size, i));

	for(int i = sizeof(original_file_size) ; i >= 1 ; i--)
		res.push_back(Hex::get_byte(original_file_size, i));

	for(int i = sizeof(packet_number) ; i >= 1 ; i--)
		res.push_back(Hex::get_byte(packet_number, i));

	for(int i = sizeof(total_packets) ; i >= 1 ; i--)
		res.push_back(Hex::get_byte(total_packets, i));

	for(int i = 0 ; i < Constants::Sizes_In_Bytes::FILE_NAME ; i++)
		res.push_back(file_name[i]);

	// we send the content after.

	return res;
}

ValidCRC::ValidCRC(const User& u) {
	this->header.init(u.uuid, Constants::Requests::codes::ValidCRC, Constants::Requests::payload_sizes::ValidCRC);
	std::copy(u.file_name, u.file_name + Constants::Sizes_In_Bytes::FILE_NAME, this->file_name);
}

std::vector<uint8_t> ValidCRC::pack() {
	std::vector<uint8_t> res = header.pack();
	for(int i = 0; i < Constants::Sizes_In_Bytes::FILE_NAME; i++)
		res.push_back(file_name[i]);

	return res;
}

InvalidCRC::InvalidCRC(const User& u) {
	this->header.init(u.uuid, Constants::Requests::codes::InvalidCRC, Constants::Requests::payload_sizes::InvalidCRC);
	std::copy(u.file_name, u.file_name + Constants::Sizes_In_Bytes::FILE_NAME, this->file_name);
}

std::vector<uint8_t> InvalidCRC::pack() {
	std::vector<uint8_t> res = header.pack();
	for(int i = 0; i < Constants::Sizes_In_Bytes::FILE_NAME; i++)
		res.push_back(file_name[i]);

	return res;
}

InvalidCRCFourthTime::InvalidCRCFourthTime(const User& u) {
	this->header.init(u.uuid, Constants::Requests::codes::InvalidCRCFourthTime, Constants::Requests::payload_sizes::InvalidCRCFourthTime);
	std::copy(u.file_name, u.file_name + Constants::Sizes_In_Bytes::FILE_NAME, this->file_name);
}

std::vector<uint8_t> InvalidCRCFourthTime::pack() {
	std::vector<uint8_t> res = header.pack();
	for(int i = 0; i < Constants::Sizes_In_Bytes::FILE_NAME; i++)
		res.push_back(file_name[i]);

	return res;
}